File: backend/src/app/utils/time_utils.py
from datetime import datetime, timezone, timedelta
from zoneinfo import ZoneInfo

IST = ZoneInfo("Asia/Kolkata")


def to_ist(dt: datetime) -> datetime:
    if dt.tzinfo is None:
        dt = dt.replace(tzinfo=timezone.utc)
    return dt.astimezone(IST)


def format_header_time(dt: datetime) -> str:
    """Format header time according to rules:
    - < 1 minute: "just now"
    - < 1 hour: "X min ago"
    - < 6 hours: "X hour(s) ago"
    - >= 6 hours and same day: "HH:MM AM/PM IST"
    - different day: "DD Mon YYYY"
    """
    now = datetime.now(timezone.utc).astimezone(IST)
    dt_ist = to_ist(dt)
    delta = now - dt_ist
    secs = delta.total_seconds()

    if secs < 60:
        return "just now"
    if secs < 3600:
        mins = int(secs // 60)
        return f"{mins} min ago"
    if secs < 3600 * 6:
        hours = int(secs // 3600)
        return f"{hours} hour{'s' if hours > 1 else ''} ago"

    # After 6 hours: if same day, show time; else show date
    if dt_ist.date() == now.date():
        return dt_ist.strftime("%I:%M %p IST")
    return dt_ist.strftime("%d %b %Y")


def format_relative_time(dt: datetime) -> str:
    """More general relative formatter for read_by timestamps (e.g., '5 min ago' or time)"""
    return format_header_time(dt)

File: frontend/src/lib/timeUtils.ts
function normalizeToISO(iso?: string | null): string | null {
  if (!iso) return null;
  // If the string already has timezone info (Z or ±HH:MM), return as-is
  if (/[Zz]$/.test(iso) || /[+-]\d{2}:\d{2}$/.test(iso)) return iso;
  // If it looks like a naive YYYY-MM-DDTHH:MM:SS, treat it as UTC
  if (/^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}(?:\.\d+)?$/.test(iso)) return `${iso}Z`;
  return iso;
}

export function formatHeaderTime(isoString?: string | null): string {
  const normalized = normalizeToISO(isoString);
  if (!normalized) return '';

  // Use Kolkata timezone for both the reference "now" and the message datetime
  const dt = new Date(normalized);
  const now = new Date();
  const dtKolkata = new Date(dt.toLocaleString('en-US', { timeZone: 'Asia/Kolkata' }));
  const nowKolkata = new Date(now.toLocaleString('en-US', { timeZone: 'Asia/Kolkata' }));

  const diffMs = nowKolkata.getTime() - dtKolkata.getTime();
  const seconds = Math.floor(diffMs / 1000);
  if (seconds < 60) return 'just now';
  if (seconds < 3600) {
    const mins = Math.floor(seconds / 60);
    return `${mins} min ago`;
  }
  if (seconds < 3600 * 6) {
    const hours = Math.floor(seconds / 3600);
    return `${hours} hour${hours > 1 ? 's' : ''} ago`;
  }

  // After 6 hours: show time in IST if same day, else date
  const dtOptions: Intl.DateTimeFormatOptions = { hour: '2-digit', minute: '2-digit', hour12: true };
  if (nowKolkata.toDateString() === dtKolkata.toDateString()) {
    return dtKolkata.toLocaleTimeString('en-US', dtOptions) + ' IST';
  }
  const dateOptions: Intl.DateTimeFormatOptions = { month: 'short', day: 'numeric' };
  return dtKolkata.toLocaleDateString('en-US', dateOptions);
}

export function formatTimeShort(isoString?: string | null): string {
  const normalized = normalizeToISO(isoString);
  if (!normalized) return '';
  const dt = new Date(normalized);
  const dtKolkata = new Date(dt.toLocaleString('en-US', { timeZone: 'Asia/Kolkata' }));
  const nowKolkata = new Date(new Date().toLocaleString('en-US', { timeZone: 'Asia/Kolkata' }));

  // If same day, show time, else show date
  if (nowKolkata.toDateString() === dtKolkata.toDateString()) {
    return dtKolkata.toLocaleTimeString('en-US', { hour: '2-digit', minute: '2-digit', hour12: true }) + ' IST';
  }
  return dtKolkata.toLocaleDateString('en-US', { month: 'short', day: 'numeric' });
}
